File: app/src/main/java/cz/marvincz/rssnotifier/room/Dao.java
package cz.marvincz.rssnotifier.room;

import android.net.Uri;

import java.util.Collection;
import java.util.List;

import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;
import cz.marvincz.rssnotifier.model.ChannelWithItems;
import cz.marvincz.rssnotifier.model.RssChannel;
import cz.marvincz.rssnotifier.model.RssItem;

@androidx.room.Dao
public abstract class Dao {
    @Query("SELECT * FROM RssChannel")
    @Transaction
    public abstract List<ChannelWithItems> getChannels();

    @Query("SELECT * FROM RssChannel WHERE RssChannel.accessUrl = :url")
    @Transaction
    abstract ChannelWithItems getChannelWithItems(Uri url);

    @Query("SELECT 1 FROM RssChannel WHERE RssChannel.accessUrl = :url")
    abstract Integer getChannelExists(Uri url);

    @Insert
    public abstract void insertChannel(RssChannel channel);

    @Delete
    @Transaction
    public abstract void deleteChannel(RssChannel channel);

    @Insert
    public abstract void insertItems(RssItem... items);

    @Insert
    public abstract void insertItems(Collection<RssItem> items);

    @Update
    public abstract void updateItems(RssItem... items);

    @Update
    public abstract void updateItems(Collection<RssItem> items);

    @Delete
    public abstract void deleteItems(RssItem... items);

    @Delete
    public abstract void deleteItems(Collection<RssItem> items);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    abstract void insertOrUpdate(Collection<RssItem> channels);

    @Transaction
    public ChannelWithItems saveChannel(RssChannel channel) {
        insertChannel(channel);
        return getChannelWithItems(channel.link);
    }

    @Transaction
    public boolean channelExists(Uri url) {
        return getChannelExists(url) != null;
    }

    @Transaction
    public List<ChannelWithItems> insertUpdateAndReturn(Collection<RssItem> items) {
        insertOrUpdate(items);
        return getChannels();
    }
}

File: app/src/main/java/cz/marvincz/rssnotifier/room/Dao.kt
package cz.marvincz.rssnotifier.room

import androidx.lifecycle.LiveData
import androidx.room.*
import cz.marvincz.rssnotifier.model.RssChannel
import cz.marvincz.rssnotifier.model.RssItem


@androidx.room.Dao
abstract class Dao {
    @Query("SELECT * FROM RssChannel ORDER BY sortOrder")
    abstract fun getChannelsLive(): LiveData<List<RssChannel>>

    @Query("SELECT * FROM RssChannel ORDER BY sortOrder")
    abstract suspend fun getChannels(): List<RssChannel>

    @Query("SELECT * FROM RssItem WHERE channelUrl = :channelUrl AND (seen = 0 OR :showSeen = 1)")
    abstract fun getItemsLive(channelUrl: String, showSeen: Boolean): LiveData<List<RssItem>>

    @Update
    abstract suspend fun updateItem(item: RssItem)

    @Query("SELECT * FROM RssChannel WHERE accessUrl = :channelUrl")
    abstract suspend fun getChannel(channelUrl: String): RssChannel?

    @Query("SELECT * FROM RssItem WHERE channelUrl = :channelUrl")
    abstract suspend fun getItems(channelUrl: String): List<RssItem>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    protected abstract suspend fun insertOrUpdate(items: Collection<RssItem>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    protected abstract suspend fun insertOrUpdate(channel: RssChannel)

    @Transaction
    open suspend fun insertOrUpdate(channel: RssChannel, items: Collection<RssItem>) {
        insertOrUpdate(channel)
        deleteChannelItems(channel.accessUrl)
        insertOrUpdate(items)
    }

    @Query("DELETE FROM RssItem WHERE channelUrl = :channelUrl")
    protected abstract suspend fun deleteChannelItems(channelUrl: String)

    @Delete
    abstract suspend fun deleteChannel(channel: RssChannel)

    @Query("SELECT MAX(sortOrder) FROM RssChannel")
    protected abstract suspend fun lastChannelOrder(): Int?

    /**
     * Order value for a new channel, so it is placed at the end
     */
    @Transaction
    open suspend fun newChannelOrder(): Int {
        return 1 + (lastChannelOrder() ?: 0)
    }

    @Update
    abstract suspend fun update(channels: Collection<RssChannel>)

    @Query("UPDATE RssItem SET seen = 1 WHERE channelUrl = :channelUrl")
    abstract suspend fun markAllRead(channelUrl: String)

    @Query("UPDATE RssItem SET seen = 1 WHERE id = :itemId")
    abstract suspend fun markRead(itemId: String)
}
